operators/pkg/controller/elasticsearch/driver/nodes.go

```diff
@@ -0,0 +1,134 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/keystore"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/reconciler"
+	esclient "github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/client"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/nodespec"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/observer"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/reconcile"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/settings"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/version/zen1"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/version/zen2"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+)
+
+func (d *defaultDriver) reconcileNodeSpecs(
+	esReachable bool,
+	esClient esclient.Client,
+	reconcileState *reconcile.State,
+	observedState observer.State,
+	resourcesState reconcile.ResourcesState,
+	keystoreResources *keystore.Resources,
+) *reconciler.Results {
+	results := &reconciler.Results{}
+
+	actualStatefulSets, err := sset.RetrieveActualStatefulSets(d.Client, k8s.ExtractNamespacedName(&d.ES))
+	if err != nil {
+		return results.WithError(err)
+	}
+
+	if !d.Expectations.GenerationExpected(actualStatefulSets.ObjectMetas()...) {
+		// Our cache of StatefulSets is out of date compared to previous reconciliation operations.
+		// This will probably lead to conflicting sset updates (which is ok), but also to
```

Comment:
Wording: you mean this 'would' probably lead to conflicting sset updates, because we are aborting the reconciliation attempt?

Reply:
I'll update the docs to make things clearer.

Reply:
```go
if !d.Expectations.GenerationExpected(actualStatefulSets.ObjectMetas()...) {
		// Our cache of StatefulSets is out of date compared to previous reconciliation operations.
		// Continuing with the reconciliation at this point may lead to:
		// - errors on rejected sset updates (conflict since cached resource out of date): that's ok
		// - calling ES orchestration settings (zen1/zen2/allocation excludes) with wrong assumptions: that's not ok
		// Hence we choose to abort the reconciliation early: will run again later with an updated cache.
		log.V(1).Info("StatefulSet cache out-of-date, re-queueing", "namespace", d.ES.Namespace, "es_name", d.ES.Name)
		return results.WithResult(defaultRequeue)
	}
```